### PR TITLE
Add SealedData wrapper for `sgx_sealed_data_t`

### PR DESCRIPTION
**Pull request by nick-mobilecoin**

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #125** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/125" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/125?utm_source=stack-comment).

**Review by jcape — CHANGES_REQUESTED** [review]

File: tservice/types/src/seal.rs [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+    pub fn key_request(&self) -> KeyRequest {
+        let bytes: [u8; 512] = self.bytes[..512]
+            .try_into()
+            .expect("SealedData bytes aren't big enough to hold `key_request`");
+        KeyRequest::from(&bytes)
+    }
+
+    // The offset within the `aes_data` for where the MAC starts.
+    // None when there is no MAC data.
+    fn mac_offset(&self) -> Option<usize> {
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply tservice/types/src/seal.rs:70]: probably a mix of bad naming on my part and questionable naming in the C api, https://github.com/intel/linux-sgx/blob/master/common/inc/sgx_tseal.h#L59

They seem to use "MAC text" a lot, and then in  a few places use [aad](https://github.com/intel/linux-sgx/blob/master/common/inc/sgx_tseal.h#L195)
My vote would be to deviate from the API and use AAD everywhere with something  in the readme or similar to let people know the intended disambiguation.

I would also like to standardize on `size` vs `len`.  The API uses both terms.  It may just be me and my history or it may be general C and embedded practice, but my understanding is `size` == number of bytes and `len` == number of items.  The type is a uint8 or u8, but with them throwing around terms like `text` it would be unfortunate for someone to mix that up with a UTF-8 String

File: core/types/src/key_request.rs [hunk 2]
```diff
@@ -72,10 +72,82 @@ bitflags! {
...
 }
 
+impl From<&[u8; 512]> for KeyRequest {
+    // Using `512` here instead of `sgx_key_request_t` to avoid leaking the
+    // abstraction to users
+    fn from(bytes: &[u8; 512]) -> Self {
+        // Unfortunately the `sgx_key_request_t` is not packed, as such we're
+        // relying on compiler convention as an assurance that the expected size
+        // is 512 bytes.
+        static_assertions::assert_eq_size!([u8; 512], sgx_key_request_t);
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply core/types/src/key_request.rs:87]: After working on #126 and #127, I'm thinking this PR/part of the implementation should change.
I'm thinking `SealedData` is generic over an `AsRef<[u8]>` and that's more or less it.  It's an opaque structure that can be passed around to the C interface, and nothing else.  TBH I'm not sure why it isn't an opaque structure in the C api.

I stayed away from the TryFrom since 512 doesn't account for the payload, but I think I better understand the payload and mac(aad) offset enough to say that we should be able to implement a tryfrom

File: core/build/src/lib.rs [hunk 3]
```diff
@@ -168,26 +168,28 @@ impl ParseCallbacks for SgxParseCallbacks {
...
     fn add_derives(&self, name: &str) -> Vec<String> {
-        if self.dynamically_sized_types.iter().any(|n| *n == name) {
-            // For dynamically sized types we don't even derive Debug, because
-            // they are often times packed and packed types can't derive Debug
-            // without deriving Copy, however by the dynamic nature one can't
-            // derive Copy
-            return vec![];
-        }
-        let mut attributes = vec!["Debug"];
+        let mut attributes = vec![];
```
Comment (awygle) [inline core/build/src/lib.rs:171]: This is just a comment not a request for changes but in general I think it's better to make changes to the build utilities in a separate commit if not a separate pull request, even if they're motivated by the "main" contents of the PR.

Comment (nick-mobilecoin) [reply core/build/src/lib.rs:171]: That sounds reasonable. With the stacking provided by graphite it should be easier to break them out and stack them appropriately.

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #125** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/125" target="_blank">Graphite</a>  👈
    * **PR #126** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/126" target="_blank">Graphite</a> 
      * **PR #127** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/127" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/125?utm_source=stack-comment).

File: core/types/src/key_request.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply core/types/src/key_request.rs]: implemented tryfrom

File: tservice/types/src/seal.rs [hunk 4]
```diff
@@ -0,0 +1,316 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+//! Types used for sealing and unsealing of secrets
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply tservice/types/src/seal.rs:3]: done

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #125** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/125" target="_blank">Graphite</a>  👈
    * **PR #126** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/126" target="_blank">Graphite</a> 
      * **PR #127** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/127" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/125?utm_source=stack-comment).

File: tservice/types/src/seal.rs [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+        let size = u32::from_le_bytes(
+            self.bytes[..4]
+                .try_into()
+                .expect("Failed to extract `payload_size`"),
+        );
+        size as usize
+    }
+
+    /// The GMAC tag of the payload
+    pub fn gmac_tag(&self) -> [u8; 16] {
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply tservice/types/src/seal.rs]: removed now

File: tservice/types/src/seal.rs [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        self.bytes[16..32]
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply tservice/types/src/seal.rs]: any left over values are computed or named constants

File: tservice/types/src/seal.rs [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+pub struct SealedData<'a> {
+    bytes: &'a [u8],
+}
+
+impl<'a> SealedData<'a> {
+    /// The key request used to seal the data
+    pub fn key_request(&self) -> KeyRequest {
+        let bytes: [u8; 512] = self.bytes[..512]
+            .try_into()
+            .expect("SealedData bytes aren't big enough to hold `key_request`");
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply tservice/types/src/seal.rs]: converted to `get()`

File: tservice/types/src/seal.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply tservice/types/src/seal.rs]: function removed now

File: core/build/src/lib.rs (hunk 3 above)
Comment (jcape) [reply core/build/src/lib.rs:171]: If you do split this out, could you make all these tests use `.contains()` instead of `.iter().any()`? That feels like something a future clippy will complain about :-)

**Review by jcape — APPROVED** [review]
LGTM

**Review by awygle — APPROVED** [review]
LGTM
